lib/spack/spack/solver/asp.py

```diff
@@ -3966,22 +3967,29 @@ def selected_specs(self) -> List[spack.spec.Spec]:
         return [s for s in self.factory() if self.is_selected(s)]
 
     @staticmethod
-    def from_store(configuration, include, exclude) -> "SpecFilter":
+    def from_store(configuration, *, include, exclude) -> "SpecFilter":
         """Constructs a filter that takes the specs from the current store."""
         packages = _external_config_with_implicit_externals(configuration)
         is_reusable = functools.partial(_is_reusable, packages=packages, local=True)
         factory = functools.partial(_specs_from_store, configuration=configuration)
         return SpecFilter(factory=factory, is_usable=is_reusable, include=include, exclude=exclude)
 
     @staticmethod
-    def from_buildcache(configuration, include, exclude) -> "SpecFilter":
+    def from_buildcache(configuration, *, include, exclude) -> "SpecFilter":
         """Constructs a filter that takes the specs from the configured buildcaches."""
         packages = _external_config_with_implicit_externals(configuration)
         is_reusable = functools.partial(_is_reusable, packages=packages, local=False)
         return SpecFilter(
             factory=_specs_from_mirror, is_usable=is_reusable, include=include, exclude=exclude
         )
 
+    @staticmethod
+    def from_environment(configuration, *, include, exclude, path=None) -> "SpecFilter":
+        factory = functools.partial(_specs_from_environment, path=path)
+        return SpecFilter(
+            factory=factory, is_usable=lambda _: True, include=include, exclude=exclude
```

Comment:
Hmm, yeah I think you are right. I changed that because I wasn't getting reuse at one point but I think the reason was related to the spec list I was passing not the `is_reusable` function.

Reply:
So I found the reason. The mock tests don't have a runtime environment, so this was failing to consider specs for reuse.